keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1535,6 +1553,10 @@ type ResourceClaimParameters struct {
     // to some unknown type.
     GeneratedFrom *ResourceClaimParametersReference
 
+    ResourceClaimParametersSpec
```

Comment:
This may be unrelated to this PR, but what happens to existing Pods if users later change the fields in the ResourceClaimParameter's embedded spec like selector? Just ignore it and keep running as is? I understand that ResourceClaim is immutable, so inline parameters cannot be changed.

Reply:
That's a good point: for the sake of feature parity, changing the `ResourceClaim.Parameters` content should be allowed.

In the current design, parameters matter right until allocation happens. After that, the parameters are ignored.

With classic DRA, a DRA driver could have watched parameters after allocation and somehow adapt the allocation result on-the-fly internally. This is not part of "structured parameters" at the moment. It's also not clear what that would mean for real-world hardware, so it's low priority.

Reply:
Actually, the ResourceClaimSpec is already mutable. That's because even if it wasn't (i.e. ResourceClassName and ParametersRef don't change after creation), that still wouldn't prevent someone from changing the content of those objects or, if they were immutable themselves, delete and recreate them.

What is protected against undesirable changes is the status. In particular, the allocation result is immutable.

Reply:
> That's because even if it wasn't (i.e. ResourceClassName and ParametersRef don't change after creation), that still wouldn't prevent someone from changing the content of those objects or, if they were immutable themselves, delete and recreate them.

I see, indeed, that perspective makes sense. It might be possible to protect deletions by setting the finalizer (`dra.k8s.io/delete-protection`) not only on the ResourceClaim but also on the ResourceClaimParameters referenced by ResourceClaim.

Anyway, for now, it seems better to make the ResourceClaimParametersSpec immutable if the allocation result of ResourceClaim cannot be changed and changes to already allocated resources are ignored. I think we should avoid behavior that confuses users. If a use case arises where users want to modify allocated resources by changing parameters, we can simply change the field from immutable to mutable, as shown in https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/storage/types.go#L287.

Reply:
>  It might be possible to protect deletions by setting the finalizer (dra.k8s.io/delete-protection) not only on the ResourceClaim but also on the ResourceClaimParameters referenced by ResourceClaim.

I'm not sure who should be doing this and when it's safe to remove the finalizer again. A finalizer also doesn't make the content immutable.

> it seems better to make the ResourceClaimParametersSpec immutable if the allocation result of ResourceClaim cannot be changed and changes to already allocated resources are ignored

This seems like a valid usage:
- create claim with parameters
- create a pod using the claim -> allocated
- change parameters
- kill pod to get it started again -> deallocated (because allocations get removed as soon as the claim is unused)
- recreate pod -> allocated with new parameters

From a conceptual perspective it seems odd to me to treat inline parameters differently from referenced parameters.

Reply:
> From a conceptual perspective it seems odd to me to treat inline parameters differently from referenced parameters.

Yes, I thought that referenced ResourceClaimParameters should be made immutable. However, I was convinced by the use cases mentioned above. It's certainly cumbersome to have to recreate ResourceClaim or ResourceClaimParameters 😄 
Currently, ResourceClaimSpec is immutable, but only inline parameters will be mutable, right?
https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/resource/validation/validation.go#L112